src/Workspaces/Core/Portable/Rename/Renamer.RenameDocumentActionSet.cs

```diff
@@ -0,0 +1,128 @@
+        /// of the actions by calling <see cref="UpdateSolutionAsync(Solution, ImmutableArray{RenameDocumentAction}, CancellationToken)"/>. 
+        /// Each action has a description of the changes that it will apply that can be presented to a user.
+        /// </summary>
+        public sealed class RenameDocumentActionSet
+        {
+            private readonly DocumentId _documentId;
+            private readonly string _documentName;
+            private readonly IReadOnlyList<string> _documentFolders;
+            private readonly ProjectId _projectId;
+            private readonly OptionSet _optionSet;
+
+            internal RenameDocumentActionSet(
+                ImmutableArray<RenameDocumentAction> actions,
+                ProjectId projectId,
+                DocumentId documentId,
+                string documentName,
+                IReadOnlyList<string> documentFolders,
+                OptionSet optionSet)
+            {
+                ApplicableActions = actions;
+                _documentFolders = documentFolders;
+                _documentId = documentId;
+                _documentName = documentName;
+                _projectId = projectId;
+                _optionSet = optionSet;
+            }
+
+            /// <summary>
+            /// All applicable actions computed for the action
+            /// </summary>
+            public ImmutableArray<RenameDocumentAction> ApplicableActions { get; }
+
+            /// <summary>
+            /// Same as calling <see cref="UpdateSolutionAsync(Solution, ImmutableArray{RenameDocumentAction}, CancellationToken)"/> with 
+            /// <see cref="ApplicableActions"/> as the argument
+            /// </summary>
+            public Task<Solution> UpdateSolutionAsync(Solution solution, CancellationToken cancellationToken)
+                => UpdateSolutionAsync(solution, ApplicableActions, cancellationToken);
+
+            /// <summary>
+            /// Applies each <see cref="RenameDocumentAction"/> in order and returns the final solution. 
+            /// All actions must be contained in <see cref="ApplicableActions" />
+            /// </summary>
+            /// <remarks>
+            /// An empty action set is still allowed and will return a modified solution
+            /// that will update the document properties as appropriate.
+            /// </remarks>
+            public async Task<Solution> UpdateSolutionAsync(Solution solution, ImmutableArray<RenameDocumentAction> actions, CancellationToken cancellationToken)
+            {
+                if (solution is null)
+                {
+                    throw new ArgumentNullException(nameof(solution));
+                }
+
+                if (actions.Any(a => !ApplicableActions.Contains(a)))
+                {
+                    throw new ArgumentException(WorkspacesResources.Cannot_apply_action_that_is_not_in_applicableactions);
```

Comment:
note, it's somewhat weird to have a localized string referencing some actual member.  this woudl make more sense as `cannot apply action that is not in '{0}'` and then use `nameof(ApplicableActions)`